### PR TITLE
Unlifted Data through `Strict`

### PR DESCRIPTION
**Pull request by sgraf812**
[Rendered](https://github.com/sgraf812/ghc-proposals/blob/unlifted-strict/proposals/0000-unlifted-strict.rst)

### CODE REVIEW

**Comment by goldfirere** [comment]
Thanks for writing this up! I'm looking forward to the design debate around the different alternatives. I was tempted to say "Let's just have unlifted data types", but you observe that your approach here separates the surface-level language concerns from code generation. I agree, and this will simplify matters. Perhaps, in the end, your `Strict` becomes user-definable, but that's backward-compatible with this proposal. I also wonder about strict unboxed tuples, but I don't have a strong enough sense of what would be best here -- hence my anticipation of the design debate.

As a smaller note: I don't understand your `tsum` example. Syntactically, I think you're missing some parentheses in pattern-matching. But also I don't see how strictness avoids the tag check, given that there are two tags. I *do* see how strictness can eliminate a tag check, but only when there is one tag.

**Comment by sgraf812** [comment]
Thanks Richard, great to have you on board :)

> Perhaps, in the end, your `Strict` becomes user-definable

Yes, I hope so. But on the other hand, I'm not sure if we can eliminate the indirection in the general (that is, almost-newtype) case. Also inference of `Force`, which could be the goal of a future proposal, would be hard to achieve/justify for all almost-newtypes. Anyway, removing the indirection is unimportant from a semantic perspective, so actually pretending that `Strict` was user-defined as in the abstract would be a reasonable model to work with.

> I don't understand your tsum example

Yes, syntax was off, indeed I missed parentheses.

Semantically, my whole proposal hinges on  [#16970](https://gitlab.haskell.org/ghc/ghc/issues/16970), specifically the fact that all strict constructor fields (`Force`s in particular) are tagged*. Hence both the `Force` thing and its wrapped value should be properly tagged. I cleared up that dependence by mentioning it in the motivation.

* Note that we also silently assume so far that `Force` things (the dreadful box) are always tagged, which might not be true for the similar reasons that strict constructor fields aren't always tagged, I guess. Although there's no `STATIC_IND` for unlifted things, since they can't be defined at the top-level. Whatever.

**Comment by goldfirere** [comment]
I'm still not following the example, I'm afraid. I think part of my confusion is that the strict example looks like it needs to check two levels of tag: both `Force` and e.g. `Leaf`. What tag check are you presuming to eliminate?

**Comment by int-index** [comment]
I don't follow the `tsum` example either. It says:

```
data Tree a = Branch !(Tree a) !a !(Tree a)
            | Leaf

tsum :: Strict (Tree Int) -> Int
tsum (Force Leaf)           = 0
tsum (Force (Branch l x r)) = tsum l + x + tsum r
```

but this does not typecheck. `l, r :: Tree Int` while `tsum` takes `Strict (Tree Int)` as input, so you need to wrap them:

```
tsum :: Strict (Tree Int) -> Int
tsum (Force Leaf)           = 0
tsum (Force (Branch l x r)) = tsum (Force l) + x + tsum (Force r)
```

Unless there's an implicit wrapping mechanism implied here?

**Comment by AndreasPK** [comment]
> I also wonder about strict unboxed tuples, but I don't have a strong enough sense of what would be best here -- hence my anticipation of the design debate.

Having suggested strict unboxed tuples I think this is the next best, if not an equal approach.
This proposal could be considered slightly less powerful as Force is restricted to lifted types. But in practice it's a power difference that doesn't matter much.

**Comment by sgraf812** [comment]
> What tag check are you presuming to eliminate?

Your confusion probably stems from the fact that by "tag check" I don't mean the actual `switch`ing involved, but rather the [zero check for the pointer tag](https://gitlab.haskell.org/ghc/ghc/issues/16820), where in the taken branch we would first have to enter the object. This check is unnecessary when we can statically prove that the tag could never be zero. Of course we still have to dispatch on the particular tag, we just don't need to enter the thing anymore to know it. Did my changes clarify this?

> so you need to wrap them

Of course, makes sense, fixed. Thanks!

**Comment by goldfirere** [comment]
We're getting closer! The key question: what is the "zero check"? The proposal links to a ticket, but the first page of the ticket does not define a "zero check", and I'm not at all familiar with this term. Thanks!

**Comment by sgraf812** [comment]
Added an explanation with a small C-- example to the motivation.

**Comment by isovector** [comment]
At a high level, this looks good to me! But I am painfully aware about how little I know in the space of unlifted reps --- offhand can anyone point me in the direction of good reading for understanding the nitty gritty in this proposal?

**Comment by goldfirere** [comment]
Thanks very much @sgraf812! I now fully understand that example. I will try to paraphrase, both to check my understanding and lightly translate to folks unfamiliar with code generation (like me):

The blob of memory that represents a lifted datatype might be in head normal form (HNF), meaning that (at least) the outermost constructor is fully evaluated. In this case, the blob contains a *tag* that corresponds to the choice of constructor. If the datatype is not in HNF, it is a thunk. This thunk will have 0s in its lower 3 bits, as it is a pointer to the code that represents evaluating the thunk. (We thus have an invariant that no tag can be divisible by 8, as 0 in the lowest three bits means "thunk".) When case-matching on a datatype, even before we look at which constructor we have, we must check whether we have any constructor at all -- in other words, we must evaluate before scrutinizing. The check for whether an object is evaluated or not is called the "zero tag check". It can be skipped if we know, statically, that an object is evaluated in HNF.

Is this summary correct? The rest of my post assumes it is.

A few questions:
* Is it possible to do the zero-tag check and scrutinize at the same time? In C, this would look something like

```c
void match(byte* obj)
{
  switch(*obj)
  {
    case 0: (*obj)();   // evaluate thunk
       match(obj); // try again
       break;
    case 1: // first constructor
    case 2: // second constructor
    ...
  }
}
```

The zero tag check is on even footing as all the other ones. Hopefully a backend can produce code that implements the check in terms of, say, local jumps and needn't do a linear search to find the matching number.

If this is done or can be done, then eliminating the zero tag check for a datatype with multiple constructor buys very little.

* Even if we can't do the zero tag check and scrutinize in the same step, it seems possible to take something like `tsum` and make a separate function `tsum_strict` that is used for the recursive occurrences. The `tsum_strict` code could certainly skip the zero tag check. Do we do this now? Would it be hard to do? If we already do this (or could easily), then the change to `tsum` saves only the initial check -- still perhaps worthwhile, but less convincingly so. (Maybe this is the worker/wrapper transformation you mention, but it doesn't require inlining unboxed sums, so I think it's different.)

-------------

I should say that I like the idea of `Strict` even if `tsum` isn't a great motivating example (it might be -- I'm still learning about this and really don't know what I'm talking about). But I do want to make sure we're clear on what this will and won't do for us.

**Comment by goldfirere** [comment]
> offhand can anyone point me in the direction of good reading for understanding the nitty gritty in this proposal?

The [Levity Polymorphism](http://repository.brynmawr.edu/cgi/viewcontent.cgi?article=1065&context=compsci_pubs) paper offers Section 2, which should be an accessible introduction. There may be other, better resources, but this one first comes to mind.

**Comment by sgraf812** [comment]
Thanks, Richard!

> Is this summary correct?

Yes, although I should mention that the blob could already have been evaluated, but the pointer not tagged yet (i.e. lower bits 0). That's what happens when a shared thunk is evaluated in one code path and then scrutinised again in another: Then we have to enter the memoised value once more to find the constructor tag.

> do the zero-tag check and scrutinize at the same time

I suppose you could. I gave this a try to see what code GHC generates for an enum of size 7. This is the result:

```
           if (R1 & 7 != 0) goto c1dK; else goto c1dL;
       c1dL: // global
           call (I64[R1])(R1) returns to c1dK, args: 8, res: 8, upd: 8;
       c1dK: // global
           switch [1 .. 7] (R1 & 7) {
               case 1 : goto c1dO;
               case 2 : goto c1dP;
               case 3 : goto c1dQ;
               case 4 : goto c1dR;
               case 5 : goto c1dS;
               case 6 : goto c1dT;
               case 7 : goto c1dU;
           }
```

So this does one extra comparison and conditional jump for the zero case, and in that case dispatches over the same jump table. We could probably get by without those additional two instructions and a backwards jump instead (punishing the presumably cold path)! I'm not sure why we aren't, but I'm not terribly invested in the back end. I'd expect LLVM to optimise this redundancy away.

> eliminating the zero tag check for a datatype with multiple constructor buys very little

Well, there's a lot of dead code and unnecessary stack movement (see the "low-level" example) which would go away completely. This probably matters for hot loops. @AndreasPK did some experiments on `containers` and found improvements of 10% on `Map.member`ship tests.

> make a separate function tsum_strict that is used for the recursive occurrences

Yes, I've thought about that a little. Let's pick up that thought once more.

It sounds a lot like constructor specialisation, only that we specialise for strictness/taggedness now, the `Force` constructor, so to speak. Or even better: We could do this as part of worker/wrapper, i.e. when we found out that a function is strict in a parameter that can't be unboxed, emit a `seq` on that parameter in the wrapper. That wrapper is then inlined at call sites and hopefully cancels away with existing statically visible evaluation, like in the recursive case of `tsum`. It's a little hard to do this without types guiding the transformation (after this proposal we could give the worker `Strict` parameters), though: How do we communicate to the worker that this particular argument is properly tagged? That's what we need `Strict` for.

> Maybe this is the worker/wrapper transformation you mention

Yes, that's exactly the kind of transformation I had in mind! But we need some means of communicating the analysis info. Worse: If we had the information available and did the specialisation, and  through some program transformation called `tsum_strict` with unevaluated arguments (think: regression in the simplifier), this wouldn't break in CoreLint at compile time but result in a segfault instead.

**Comment by AndreasPK** [comment]
> do the zero-tag check and scrutinize at the same time:

Jump tables are generally worse if we only have few possible targets, so your code snippet is slightly misleading in terms of performance.
The current code we generate is in fact a compiled representation of such a switch, but optimized for the unevaluated case. There are reasons for doing this but they are beyond this proposal. See [here](https://gitlab.haskell.org/ghc/ghc/issues/8905) for some discussion.

But either way the issue of zero-tag checks exists even for types with a single Constructor.

We might, for example, want to take out the second field of a Constructor like a tuple. If we know we are given a pointer to the constructor itself the Cmm code will be:

```C
           R1 = P64[R1 + 7]; //Offset of field - tag
           call (I64[R1])(R1) //"return R1"
```

However currently we have to check if the argument is evaluated first so we get instead:

```C
//Set up the stack, check the tag, evaluate argument if needed.
       c1dr: // global
           if ((Sp + -8) < SpLim) (likely: False) goto c1dv; else goto c1dw;
       c1dv: // global
           R2 = R2;
           R1 = Foo.foo_closure;
           call (stg_gc_fun)(R2, R1) args: 8, res: 0, upd: 8;
       c1dw: // global
           I64[Sp - 8] = c1do;
           R1 = R2;
           Sp = Sp - 8;
           if (R1 & 7 != 0) goto c1do; else goto c1dp;
       c1dp: // global
           call (I64[R1])(R1) returns to c1do, args: 8, res: 8, upd: 8;
       c1do: // global
           I64[Sp] = c1du;
//Finally, get the field and return it.
           R1 = P64[R1 + 7];
           call (I64[R1])(R1) //"return R1"
```

The cost here is not simply the branch, but we also spill to the stack, adjust the stack pointer, potentially spill live variables. If other code does not require spills to the stack it will also incur another branch for a stack check.

There is no point in pretending that this will be a huge improvement for a typical haskell program. But there is a small number of code patterns which do suffer from this overhead and they will benefit greatly.

> Even if we can't do the zero tag check and scrutinize in the same step, it seems possible to take something like tsum and make a separate function tsum_strict that is used for the recursive occurrences. The tsum_strict code could certainly skip the zero tag check. Do we do this now?

In short no. 

@sgraf812 answered most of this already but to be explicit there is currently no way to define tsum_strict this way. Users can't define it this way.
The compiler won't try to transform it into something like this either, primarily because GHC does internally have no notion of "boxed but guaranteed unlifted" as Force would provide/require.

This proposal would allow users to write such functions by hand, and make transformations towards this goal in the compiler easier.

**Comment by goldfirere** [comment]
Thanks @sgraf812 and @AndreasPK! This all makes much more sense to me now. The key bit for me was about the fact that there would be no way to communicate to `tsum_strict` not to evaluate its argument. Somehow I thought there was a mechanism for that, but clearly there isn't. And indeed `Strict` provides precisely that mechanism.

My support for this proposal has grown stronger. Many thanks for taking the time to educate me!

**Comment by AndreasPK** [comment]
In 3.3 some trailing code seems to have snuck into the code block.

**Comment by simonpj** [comment]
I'm interested in discussing the tradeoff between

1. A single unlifted `Force` type as you propose
2. Allowing users to declare new unlifted types, as [proposed here](https://gitlab.haskell.org/ghc/ghc/wikis/unlifted-data-types#proposal-1-allow-data-types-to-be-declared-as-unlifted).

Advantages of (1)

* One new built-in primitive type
* You can make an existing type strict by wrapping it in `Force`

Disadvantages of (1)

* All this pattern-matching on `Force` and use of `Force` as a constructor (which you can see in `tsum`) adds a lot of clutter to the source code.  Indeed, in the defn of `Ptr` in 3.1, the `Addr#` is wrapped in no fewer than _three_ data constructors: `Ptr`, `Force` and `Ptr#`!

* All this pattern-matching on `Force` and use of `Force` as a constructor may add runtime cost.  In particular, a data type like `Strict` alwasya allocates a box, and that is expensive -- which is particularly unacceptable in the performance critical code for which this feature is intended.

  Alternatively, perhaps we can arrange that `Force` is implemented by nothing at all, a bit like `coerce`??   But that really would be entirely new.   And this "coercion" would be between types of kind `TYPE LiftedRep` and `TYPE UnliftedRep` respectively, and generally we don't let you coerce between types with different runtime-representations.

Advantages of (2)

* All the additional pattern matching and construction goes away.  For example, we can just say
  ```
  data unlifted Ptr a = Ptr Addr#
  ```
  and we are done.

* `Strict` is definiable as a special case.  (But it will be boxed unless we do more work.)

Disadvantages of (2) (maybe)

* The proposal says "This would have to anticipate all possible interactions with existing ways to introduce (generalised) algebraic data types, from surface language issues regarding syntax to code generation issues".   I don't see the problem here.  We'd just need the new user-defined unlifted data type to have kind `TYPE UnliftedRep`.   _We already have type constructors of this kind_ so nothing new is needed to deal with them.  I think it might "just work".

To me (2) seems rather attractive.

**Comment by sgraf812** [comment]
I started implementing a `Strict#` primtype with `mkForce#`/`getForce#` pseudoops and noticed that it's quite hard to come up with all the builtin rewrite rules, when all we are basically saying is that `mkForce#` and `getForce#` are `DataCon`s. Implementing `Strict` as a data constructor is probably the proper solution here, at which point we could just give the user the additional bit of syntax (well, omission of semantic check, rather) to declare them themselves. I'm a little wary about unforeseen interactions, but you seem pretty convinced that the hard bits should already be hammered out. So why not? I'll prepare a counter proposal for that.

I actually consider removal of the box essential to making `Force` a useful thing, by the way. We could apply this optimisation to any 'newtype' that turns lifted into unlifted pointers.

> Alternatively, perhaps we can arrange that Force is implemented by nothing at all, a bit like coerce?? But that really would be entirely new. And this "coercion" would be between types of kind TYPE LiftedRep and TYPE UnliftedRep respectively, and generally we don't let you coerce between types with different runtime-representations.

Yes, exactly. I'm not sure this would be covered by the current `Coercible` mechanism, because they are supposed to be zero-cost (which a `coerce` from lifted to unlifted rep isn't, it's a `seq`). But maybe we could introduce a super-class of `Coercible` for that (`CostlyCoercible` 🎉), or introduce an entirely new mechanism.

**Comment by sgraf812** [comment]
`Force` is still useful as a library solution, by the way: Strict fields with pointer representation should desugar to a `DataCon` wrapper that forces strict field arguments and a worker that actually has `Force`-wrapped arguments. Example

```
data SPair a b = T !a !b
-- introduces T :: Strict a -> Strict b -> SPair a b
-- with the following wrapper
$WPair :: a -> b -> TPair a b
$WPair x y = T (Force a) (Force b)
```

**Comment by goldfirere** [comment]
I think we can implement `Strict` without a box. We'll just need primitives

```
forceCoerce :: forall (a :: TYPE LiftedRep). a -> Strict a
   -- just forces. No box.

unpackStrict :: forall (a :: TYPE LiftedRep). Strict a -> a
  -- does nothing at all at runtime
```

Here, we take advantage of the fact that (as I understand it) the set of possible representations for a value of type of kind `TYPE UnliftedRep` is a subset of the equivalent for `TYPE LiftedRep`: that is, the former just doesn't have thunks. So as long as we know something isn't a thunk, we can just change its kind. Unpacking the lifted value out of a `Strict` is then a no-op.

**Comment by sgraf812** [comment]
> We'll just need primitives

This is basically the approach I tried implementing in https://github.com/ghc-proposals/ghc-proposals/pull/257#issuecomment-523073017. I found it unsatisfying, because I have the fear that the primitives will trip up the simplifier and other optimisations. For example, I *think* we'd need the rewrite rules `forall a. forceCoerce (unpackStrict a) = a` and ``forall v. let a = forceCoerce v in f (unpackStrict a) = v `seq` f v``. The reasoning would all come for free if `Strict` was just a `DataCon`.

I think eliminating the box is purely a CodeGen optimisation, so we probably shouldn't introduce new pseudoops that basically are just data constructors.

**Comment by treeowl** [comment]
Since I made a [similar suggestion](https://mail.haskell.org/pipermail/ghc-devs/2018-February/015356.html), I'm in favor. I do think it would be nice to guarantee that there's no box. That seems even more important than guaranteeing no zero-tag check.

**Comment by treeowl** [comment]
One gripe: every other primitive unlifted thing I know of has a hash at the end of its name. Shouldn't this too?

**Comment by sgraf812** [comment]
Here is the 'Unlifted Data' counter proposal: https://github.com/ghc-proposals/ghc-proposals/pull/265

**Comment by code5hot** [comment]
I have a need to be able to trivially define an unlifted type for any datatype with exactly one constructor, we do this for datatypes with one constructor *and* one field using `newtype` but the constraint should just be "exactly one constructor" and otherwise be semantically similar to `newtype`. I started a proposal at #268 but it's not fully refined yet.

**Comment by treeowl** [comment]
Speaking of one-way coercions: if we ever get those, we'll also want

```haskell
oneWayCoerce :: a -> (# a #)
```

which I've mentioned before. One-way coercions need to be careful about covariance and contravariance.

**Comment by sgraf812** [comment]
Closing this as superseded by #265 which already has a propotype implementation
